### PR TITLE
Fix tests and connect to CI

### LINKED ISSUE #2: Add testing on travis builds

### PR DESCRIPTION
**Pull request by Waldz**
Closes #2
Updates #14

### CODE REVIEW

File: ci/commands/command_unittests.go [hunk 1]
```diff
@@ -0,0 +1,10 @@
+package commands
+
+import "fmt"
+
+// DoUnitTests runs all unit tests in project
+func DoUnitTests(_ []string) {
+	MustRun("go", "test", "./...")
```
Comment (vkuznecovas) [inline ci/commands/command_unittests.go]: might want to include -race and -cover

File: ci/main.go [hunk 2]
```diff
@@ -0,0 +1,36 @@
...
+`
+
+	args := os.Args[1:]
+	if len(args) < 1 {
+		fmt.Print(help)
+		commands.ExitWithError("Need subcommand as first argument")
+	}
+
+	command := strings.ToLower(args[0])
+	switch command {
```
Comment (vkuznecovas) [inline ci/main.go]: this works, but won't work well looking at the future. If we add tons of scripts here, we'll need to have separate helps for them and what not. The switch will be hard to read if we add them here. At least use an urfave/cli here, or better yet - something like [magefile](https://magefile.org/).

File: .travis.yml [hunk 3]
```diff
@@ -1,27 +1,73 @@
-language: go
 sudo: true #for docker service to be enabled
 os: linux
+language: go
+go: [1.9.2]
```
[the remark this answers is not in the document]
Comment (Waldz) [reply .travis.yml]: For that we need to prepare XGO container with 1.11
Created #20, feel free to take it

File: .travis.yml [hunk 4]
```diff
@@ -1,27 +1,73 @@
...
+    #if: tag IS present
+    if: type = never
+
+
+jobs:
+  include:
+    # Dependencies cache stage
+    - stage: dep-cache
+      name: "Dependencies update"
+      script: make deps
```
[the remark this answers is not in the document]
Comment (Waldz) [reply .travis.yml]: `make` is fo people who dont understand Golang ecosystem. Or if You're tired to type `scripts/...` or `go run ci/main.go ...`

```
# This Makefile is meant to be used by people that do not usually work with Go source code.
# If you know what GOPATH is then you probably don't need to bother with make.
```

File: .travis.yml [hunk 5]
```diff
@@ -1,27 +1,73 @@
...
+    # Test stage
+    - stage: test
+      script: scripts/xgo_run.sh make test
+      name: "Unit tests"
+
+    # Build artifacts
+    - stage: build
+      script:
+      - go get github.com/karalabe/xgo
+      - ./check-all.sh $PWD/examples/desktop
```
[the remark this answers is not in the document]
Comment (Waldz) [reply .travis.yml]: Out of scope

File: scripts/xgo_run.sh [hunk 6]
```diff
@@ -0,0 +1,15 @@
...
+set -e
+
+COMMAND=$@
+if [ -z "$COMMAND" ]; then
+    printf "\e[0;31m%s\e[0m\n" "Missing command!"
+    exit 1
+fi
+
+docker run -it --rm \
+    -v `pwd`:/go-src-root/ \
```
[the remark this answers is not in the document]
Comment (Waldz) [reply scripts/xgo_run.sh]: Out of scope. Created #18 feel free to take it

File: ci/main.go [hunk 7]
```diff
@@ -0,0 +1,36 @@
+package main
+
+import (
+	"fmt"
+	"github.com/mysteriumnetwork/go-openvpn/ci/commands"
```
[the remark this answers is not in the document]
Comment (Waldz) [reply ci/main.go]: Out of scope. Created #21.

File: ci/util/cmd.go [hunk 8]
```diff
@@ -0,0 +1,25 @@
...
+// GetGoBinaryPath looks for the given binary in path, if not checks if it's in $GOPATH/bin
+func GetGoBinaryPath(binaryName string) (string, error) {
+	res, err := sh.Output("which", binaryName)
+	if err != nil {
+		return "", err
+	}
+	if res != "" {
+		return res, nil
+	}
+	gopath := os.Getenv("GOPATH")
```
Comment (soffokl) [inline ci/util/cmd.go]: `GOPATH` variable can be empty starting from 1.8 https://golang.org/doc/go1.8#gopath
In this case, it has a default value:`$HOME/go`

File: .travis.yml [hunk 9]
```diff
@@ -1,27 +1,72 @@
...
+
+    # Test stage
+    - stage: test
+      script: scripts/xgo_run.sh make test
+      name: "Unit tests"
+
+    # Build artifacts
+    - stage: build
+      script:
+      - go get github.com/karalabe/xgo
```
Comment (zolia) [inline .travis.yml]: Why not use our own: https://github.com/mysteriumnetwork/xgo ?

Comment (soffokl) [reply .travis.yml]: I believe that it's better to use an outside project if it's possible because it's maintainable.

Comment (vkuznecovas) [reply .travis.yml]: Why I agree that it's better to use an outside project, I have a feeling this was forked for a reason.

File: ci/util/cmd.go (hunk 8 above)
Comment (vkuznecovas) [reply ci/util/cmd.go]: forgot to mention - fixd

File: ci/install_dependencies.go [hunk 10]
```diff
@@ -0,0 +1,29 @@
...
+import (
+	"fmt"
+
+	"github.com/magefile/mage/mg"
+	"github.com/magefile/mage/sh"
+	"github.com/mysteriumnetwork/go-openvpn/ci/util"
+)
+
+// Installs go dependencies
+func Deps() error {
```
[the remark this answers is not in the document]
Comment (vkuznecovas) [reply ci/install_dependencies.go]: will do

**Comment by Waldz** [comment]
LGTM

File: openvpn/cmd_wrapper_test.go [hunk 11]
```diff
@@ -32,10 +33,12 @@ func TestHelperProcess(t *testing.T) {
...
 func TestWaitAndStopProcessDoesNotDeadLocks(t *testing.T) {
 	execTestHelper := NewExecCmdTestHelper("TestHelperProcess")
-	execCommand := execTestHelper.ExecCommand
+	execCommand := func(arg ...string) *exec.Cmd {
+		return execTestHelper.ExecCommand("openvpn", arg...)
+	}
 	execTestHelper.AddExecResult("", "", 0, 10000, "openvpn")
 
-	process := NewCmdWrapper("openvpn", "[process-log] ", execCommand)
+	process := NewCmdWrapper("[process-log] ", execCommand)
```
Comment (zolia) [inline openvpn/cmd_wrapper_test.go]: we could make "[process-log]" as prefix, or use global one from openvpn package.

Comment (vkuznecovas) [reply openvpn/cmd_wrapper_test.go]: moved the prefixes to constants.

File: scripts/build-bridge.sh [hunk 12]
```diff
@@ -1,16 +1,18 @@
...
 fetch_openvpn3
 
+rm -rf /go-src-root/openvpn3/bridge/*.a /go-src-root/openvpn3/bridge/*.h
 cp -f core/adapter/*.h /go-src-root/openvpn3/bridge/.
 
 export O3=`pwd`
 export DEP_DIR=`pwd`/dep_dir
 mkdir -p $DEP_DIR
-export DL=/build/dls
+export DL=`pwd`/dls
```
Comment (zolia) [inline scripts/build-bridge.sh]: we can use $PWD variable, should be slightly faster.

File: scripts/xgo_run.sh [hunk 13]
```diff
@@ -0,0 +1,15 @@
...
+    -v `pwd`:/go/src/github.com/mysteriumnetwork/go-openvpn \
+    -w /go/src/github.com/mysteriumnetwork/go-openvpn \
+    --entrypoint "/bin/bash" \
+    mysteriumnetwork/xgo:1.11 -c "${COMMAND}"
```
Comment (zolia) [inline scripts/xgo_run.sh]: We prob should have xgo:myst tag we support, so we should not go over and over all sources changing every time we bump version up. Created a ticket for this. #414

Comment (vkuznecovas) [reply scripts/xgo_run.sh]: fair point

**Pull request merged**